contrib/config/test/kv_store_xds_delegate_integration_test.cc

```diff
@@ -250,6 +250,25 @@ class KeyValueStoreXdsDelegateIntegrationTest
     // Set up a new Envoy, using the previous Envoy's configuration, and create the test server.
     ConfigHelper helper(version_, *api_,
                         MessageUtil::getJsonStringFromMessageOrError(config_helper_.bootstrap()));
+
+    // Close the connection between Envoy and the xDS FakeUpstreams.
+    closeConnection(sds_connection_);
+    closeConnection(rtds_connection_);
+    // Bring down the xDS FakeUpstreams.
+    getSdsUpstream().reset();
+    getRtdsUpstream().reset();
+
+    // Reset the xDS upstreams to new FakeUpstreams, since we'll re-establish a connection to xDS
+    // after the Envoy server gets re-created and the KV store loaded.
+    fake_upstreams_.pop_back(); // Removes the RTDS upstream.
```

Comment:
when you say explicit, do you mean adding a `removeFakeUpstream()` function to BaseIntegrationTest, as well as using `addFakeUpstream` and adding a param to specify deferred initialization?

Reply:
I think that the `fake_upstreams_` were expected to be used for backend hosts, and somehow many of the integration tests are using them also for xDS servers.
It seems that the case here needs more control over those explicitly, so it may be better to use 2 explicit "upstreams" members: `rtds_upstream_`, and `sds_upstream_` each having the FakeUpstream, its connection, etc.
That way the removal/addition to the fake_upstreams_ is avoided.

An example where the current use of `fake_upstreams_` could backfire is if there was also data-plane cluster, which was added after the 2 (rtds, sds) upstreams.

No AI here for now, just something to think about for the future.

Reply:
ah ok I see what you mean now.  That's a fair point, though I think we'll have to rework some logic in BaseIntegrationTest.  Other xDS tests follow the same pattern of using fake_upstreams_ for the xDS servers too, e.g. https://github.com/envoyproxy/envoy/blob/e2a40eaff8baf35b269a936526512417033b916a/test/integration/xds_integration_test.cc#L869-L878

I agree - I think it's worth trying to figure out a better way for this in a follow up PR.